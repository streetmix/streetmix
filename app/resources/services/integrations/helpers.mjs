import models from '../../../db/models/index.js'

const { User, UserConnections, sequelize } = models

/**
 finds the database record for the given user
 */
// might be better to do something besides return null
// we can actually probably get rid of this function if we want, but it might help
// keep some of the other code clean
export async function findUser (userId) {
  const user = await User.findOne({ where: { auth0Id: userId } })
  if (user === null) {
    return null
  }
  return user.dataValues
}

// if the provider we're trying to find exists
// return its index value or -1 if not found
export function addOrUpdateByProviderName (array, item) {
  const indexValue = array.findIndex(
    (_item) => _item.provider === item.provider
  )
  if (indexValue === -1) {
    array.push(item)
  } else {
    array[indexValue] = item
  }
}

/**
 given a user, check their identity data,
 depending on what data you find, add subscriber role

 TODO: make this less permissive(maybe add a status field or have logic depending on the provider), tested,
 and remove the role if we're confident we should (probably shouldn't worry about this for MVP)
 */
export async function syncAccountStatus (userId) {
  const databaseUser = await User.findOne({ where: { auth0Id: userId } })
  if (databaseUser.identities.some((identity) => 'user_id' in identity)) {
    databaseUser.addRole('SUBSCRIBER_1')
  }
}

/**
 * When a user has successfully connected their Streetmix account with a third-
 * party identity provider, record that connected identity with the user's
 * Streetmix data in the UserConnection table.
 *
 * @todo Remove first attempt at storing identity information as a JSON blob
 *        on a User model
 * @param {Object} account - User's Streetmix account information
 * @param {Object} profile - User's connected identity information
 * @returns {Promise}
 */
export function addUserConnection (account, profile) {
  const identities = account.identities
  const identity = {
    provider: profile.provider,
    user_id: profile.id,
    access_token: profile.access_token || undefined, // Coil
    refresh_token: profile.refresh_token || undefined // Coil
  }

  // if provider exists, should update that item
  // if not, should add to the object list
  // TODO: Deprecated
  addOrUpdateByProviderName(identities, identity)

  return sequelize.transaction(async (t) => {
    // This is a previous attempt, where we saved connected identity
    // information to a JSON field in the User model. A newer method (using
    // the UserConnection table) is used below. Both are currently used to
    // test concurrent data validity. In time, the User.update() method should
    // be removed.
    await User.update(
      {
        identities
      },
      { where: { auth0Id: account.auth0Id }, returning: true, transaction: t }
    )

    await UserConnections.findOrCreate({
      where: {
        user_id: account.id,
        provider: profile.provider
      },
      transaction: t
    })
    return await UserConnections.update(
      {
        provider_user_id: profile.id,
        metadata: profile
      },
      {
        where: {
          user_id: account.id,
          provider: profile.provider
        },
        returning: true,
        transaction: t
      }
    )
  })
}
