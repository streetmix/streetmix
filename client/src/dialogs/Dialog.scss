@import "../../styles/mixins";

.dialog-box-container {
  --dialog-border-radius: var(--border-radius-large);
  --header-background-color: var(--color-turquoise-200);
  --header-text-color: var(--color-turquoise-700);

  display: flex;
  align-items: center;
  justify-content: center;
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  left: 0;
}

.dialog-box-backdrop {
  z-index: var(--z-08-dialog-box-backdrop);
  position: absolute;
  inset: 0;
  background: var(--sky-color);
  touch-action: none;

  /* Transitions */
  transition: opacity 60ms ease-in-out;
  opacity: 0;
}

.dialog-box {
  position: relative;

  // balance the dialog by making it slightly north of center
  margin-top: -3%;
  padding: 0;
  background: white;
  line-height: 1.4;
  z-index: var(--z-09-dialog-box);
  touch-action: none;

  /* New styles! Let's try it! */
  border-radius: var(--dialog-border-radius);
  overflow: hidden;
  box-shadow: 0 0 20px 0 rgb(0 0 0 / 7.5%);

  /* Transitions */
  transition:
    opacity 80ms ease-in-out,
    transform 80ms ease-in-out;
  opacity: 0;
  transform: scale(0.9);

  /* Containing child element */
  > div {
    display: flex;
    flex-direction: column;
  }

  header {
    padding: 1.75rem 1em;
    text-align: center;
    background-color: var(--header-background-color);
    color: var(--header-text-color);
  }

  h1 {
    @include large-message-text;

    margin: 0;
  }

  footer {
    padding: 1rem 2rem;
    margin-top: 0;
    color: rgb(160 160 160);
    font-size: 0.85em;

    > p:first-child {
      margin-top: 0;
    }

    a {
      color: inherit;
    }

    a:hover {
      color: rgb(120 120 120);
    }
  }

  button.close {
    z-index: var(--z-09-dialog-box);
  }

  // Button placement
  button.btn,
  a.btn {
    margin: 0 auto;
    margin-top: 1em;
  }
}

/* Transitions */
.dialog-transition-enter-done .dialog-box {
  opacity: 1;
  transform: scale(1);
}

.dialog-transition-exit .dialog-box {
  opacity: 0;
  transform: scale(0.9);
}

.dialog-transition-enter-done .dialog-box-backdrop {
  opacity: 0.5;
}

.dialog-transition-exit .dialog-box-backdrop {
  opacity: 0;
}

.dialog-content {
  flex-grow: 1;
  position: relative;

  &:not(.dialog-content-bleed) {
    padding: 1rem 2rem;
  }

  > p:first-child {
    margin-top: 0;
  }
}

button.dialog-primary-action {
  border: 0;
  border-top: 1px solid var(--color-turquoise-350);
  border-radius: 0;
  background-color: white;
  font-weight: normal;
  width: 100%;
  text-align: center;
  color: var(--color-turquoise-600);
  min-height: 46px;
  cursor: pointer;

  /* override .dialog-box button */
  margin-top: 0 !important;

  &:hover {
    background-color: var(--color-turquoise-100);
    color: var(--color-turquoise-700);
  }
}
