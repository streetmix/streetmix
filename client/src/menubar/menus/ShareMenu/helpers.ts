import type { UserState } from '~/src/types'
import type { StreetState } from '@streetmix/types'
import type { IntlShape } from 'react-intl'

export interface SocialShareProps {
  shareText: string
  shareUrl: string
}

/**
 * Returns the current page URL.
 * This may eventually change to a short URL generator - strt.mx
 */
export function getSharingUrl (): string {
  return window.location.href
}

export function getSharingMessage (
  street: StreetState,
  user: UserState,
  intl: IntlShape
): string {
  const signedIn = user.signedIn ?? false
  const userId = user.signInData?.userId ?? ''
  let message = ''

  if (typeof street.creatorId === 'string') {
    if (signedIn && street.creatorId === userId) {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.my-street',
            defaultMessage: 'Check out my street, {streetName}, on Streetmix!'
          },
          { streetName: street.name }
        )
      } else {
        message = intl.formatMessage({
          id: 'menu.share.messages.my-street-unnamed',
          defaultMessage: 'Check out my street on Streetmix!'
        })
      }
    } else {
      if (typeof street.name === 'string') {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street',
            defaultMessage:
              'Check out {streetName} by {streetCreator} on Streetmix!'
          },
          { streetName: street.name, streetCreator: `@${street.creatorId}` }
        )
      } else {
        message = intl.formatMessage(
          {
            id: 'menu.share.messages.someone-elses-street-unnamed',
            defaultMessage:
              'Check out this street by {streetCreator} on Streetmix!'
          },
          { streetCreator: `@${street.creatorId}` }
        )
      }
    }
  } else {
    if (typeof street.name === 'string') {
      message = intl.formatMessage(
        {
          id: 'menu.share.messages.anonymous-creator-street',
          defaultMessage: 'Check out {streetName} on Streetmix!'
        },
        { streetName: street.name }
      )
    } else {
      message = intl.formatMessage({
        id: 'menu.share.messages.anonymous-creator-street-unnamed',
        defaultMessage: 'Check out this street on Streetmix!'
      })
    }
  }

  return message
}
