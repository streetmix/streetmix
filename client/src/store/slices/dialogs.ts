import { createSlice } from '@reduxjs/toolkit'
import type { PayloadAction } from '@reduxjs/toolkit'

interface DialogState {
  name: string | null
}

// It may be desirable to simplify the shape of this state object
// to just a string - see the menus state for example.
// This used to have more properties in an earlier implementation,
// and is currently selected as a prop object by the <DialogRoot>
// component, which means this can't be refactored until that is.
const initialState: DialogState = {
  name: null
}

const dialogsSlice = createSlice({
  name: 'dialogs',
  initialState,

  reducers: {
    showDialog (state, action: PayloadAction<DialogState['name']>) {
      state.name = action.payload
    },

    clearDialogs (state, action: PayloadAction<undefined>) {
      state.name = null
    }
  }
})

export const { showDialog, clearDialogs } = dialogsSlice.actions

export default dialogsSlice.reducer
