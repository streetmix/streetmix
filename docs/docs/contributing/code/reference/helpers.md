---
sidebar_position: 2
---

# Helpers and utilities

:::caution Under construction

This page is a work in progress.

:::

We use the following utilities to solve common programming problems.

## Pseudo-random number generator

If you need a one-time use random number (which will never be stored), there's nothing wrong with JavaScript's built-in `Math.random()`. However, there are times where we need a **seed** for a pseudo-random number generator (PRNG). Seeds generate a random-_looking_ sequence of numbers, but they have the benefit of remaining consistent across renders. For example, this is useful on our sidewalk segment, where we want to generate a random set of pedestrians but we want those pedestrians to remain the same between page refreshes or React component updates.

We can generate seeds by calling `generateRandSeed()`, then by feeding those seeds into the `seedrandom` library:

```js title="random-example.js"
import seedrandom from "seedrandom";
import { generateRandSeed } from "./util/random";

function doThingsWithThePRNG() {
  const seed = generateRandSeed();
  const prng1 = seedrandom(seed);

  // Generate pseudo-random numbers. Each time rng() is called, it returns
  // the next pseudo-random number in the sequence.
  console.log(prng1()); // -> first pseudo-random number
  console.log(prng1()); // -> second pseudo-random number
  // .. etc

  // Create another generator with the same seed
  const prng2 = seedrandom(seed);
  console.log(prng2()); // -> same value as the first call to prng1()
}
```

When `seedrandom()` is called, it returns a function that will return a consistent sequence of numbers each time it is called, for the given `seed`. The sequence returned is local to that function.

:::note

Although the `seed` can be any type of value, like a string, `generateRandSeed()` returns only integer values so that when we store seeds in a data model we can check for type consistency.

:::

In the above example, we call `seedrandom()` twice, with the same `seed`, so that we can create the same sequence of pseudo-random numbers in two different scenarios. This example is contrived because the pseudo-random number generators are used in the same function, and in reality the `seedrandom()` call may happen in separate parts of the application.

**References**

- [Random Seeds, Coded Hints, and Quintillions](http://davidbau.com/archives/2010/01/30/random_seeds_coded_hints_and_quintillions.html) [David Bau]
- [seedrandom.js](https://github.com/davidbau/seedrandom) [GitHub]

## axios

The [Axios](https://github.com/axios/axios) library was introduced as a recommendation over native `fetch` as a way to manage requests to the Streetmix back-end API layer. Additionally, Axios Mock Adapter offers the possibility to mock all API requests for integration testing.

If you work on API-related features, consider refactoring the `fetch` calls to use the API in `.assets/scripts/util/api.js`.
