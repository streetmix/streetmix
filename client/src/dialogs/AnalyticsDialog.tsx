/**
 * Street Analytics (dialog box)
 *
 * Renders the "Analytics" dialog box.
 *
 */
import React, { useState, useRef } from 'react'
import { FormattedMessage, useIntl } from 'react-intl'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { useSelector, useDispatch } from '../store/hooks'
import { setShowAnalytics } from '../store/actions/street'
import Terms from '../app/Terms'
import Button from '../ui/Button'
import Checkbox from '../ui/Checkbox'
import ExternalLink from '../ui/ExternalLink'
import { ICON_QUESTION_CIRCLE } from '../ui/icons'
import { isOwnedByCurrentUser } from '../streets/owner'
import { formatNumber } from '../util/number_format'
import {
  getCapacityData,
  getStreetCapacity,
  getRolledUpSegmentCapacities,
  saveCsv
} from '../segments/capacity'
import { SETTINGS_UNITS_IMPERIAL } from '../users/constants'
import CapacitySources from './Analytics/CapacitySources'
import SegmentAnalytics from './Analytics/SegmentAnalytics'
import Dialog from './Dialog'
import './AnalyticsDialog.scss'

function AnalyticsDialog (): React.ReactElement {
  const street = useSelector((state) => state.street)
  const locale = useSelector((state) => state.locale.locale)
  const dispatch = useDispatch()
  const intl = useIntl()
  const max = useRef(0)

  const [isVisible, setVisible] = useState(street.showAnalytics)
  const toggleVisible = (): void => {
    setVisible(!isVisible)
    void dispatch(setShowAnalytics(!isVisible))
  }

  const capacityData = getCapacityData(street.capacitySource)
  const capacity = getStreetCapacity(street)
  const options = { maximumSignificantDigits: 3 }

  const summary = (
    <FormattedMessage
      id="dialogs.analytics.street-summary"
      defaultMessage="Your street has an estimated average traffic of {averageTotal} people per hour, and potential for up to {potentialTotal} people per hour."
      values={{
        averageTotal: <b>{formatNumber(capacity.average, locale, options)}</b>,
        potentialTotal: (
          <b>{formatNumber(capacity.potential, locale, options)}</b>
        )
      }}
    />
  )

  // Displays typical lane width from the data source, if present. Values are
  // manually set for each unit type (metric or imperial) instead of using the
  // `prettifyWidth` helper, which has imprecise rounding. While every data
  // source we have right now does have an assumed lane width, this property
  // is optional, in case of future data sources that calculate capacity using
  // lane width as an input variable.
  let laneWidth

  if (street.units === SETTINGS_UNITS_IMPERIAL) {
    laneWidth = `${capacityData.typical_lane_width.imperial} ft`
  } else {
    laneWidth = `${capacityData.typical_lane_width.metric} m`
  }

  const widthText = (
    <FormattedMessage
      id="dialogs.analytics.typical-lane-width"
      defaultMessage="Capacity values are based on {laneWidth}-wide lanes."
      values={{ laneWidth }}
    />
  )

  const rolledUp = getRolledUpSegmentCapacities(street)

  // Store the maximum capacity across renders, when the data source may
  // change. This way, data sources with lower values show more intuitive
  // animation between the relative differences. (Note: one downside of this
  // method is that the 'max' is only calculated with data sources that are
  // being tried - we don't calculate the max across _all_ data sources right
  // away. The City of Vancouver data (with the lowest capacity numbers) will
  // display at max width, even though it would be significantly shorter if
  // you switched to the TUMI/GIZ data first and then back to Vancouver. This
  // is a tradeoff I'm willing to live with, unless user feedback really
  // requires us to chnage this.)
  max.current = Math.max(
    max.current,
    ...rolledUp.map((item) => item.capacity?.potential ?? 0)
  )

  function exportCSV (): void {
    const name =
      street.name ??
      intl.formatMessage({
        id: 'street.default-name',
        defaultMessage: 'Unnamed St'
      })
    saveCsv(rolledUp, name)
  }

  return (
    <Dialog>
      {(closeDialog) => (
        <div className="analytics-dialog">
          <header>
            <h1>
              <FormattedMessage
                id="dialogs.analytics.heading"
                defaultMessage="Analytics"
              />
            </h1>
          </header>
          <div className="dialog-content">
            <div className="analytics-dialog-content">
              <p>
                {summary} {widthText}
              </p>
              {rolledUp.map((item, index) => (
                <SegmentAnalytics
                  key={index}
                  index={index}
                  max={max.current}
                  type={item.type}
                  capacity={item.capacity}
                />
              ))}
              <p>
                <strong>
                  <FormattedMessage
                    id="dialogs.analytics.source"
                    defaultMessage="Source"
                  />
                  :
                </strong>{' '}
                {capacityData.source_url !== undefined
                  ? (
                    <ExternalLink href={capacityData.source_url}>
                      {capacityData.source_title}
                    </ExternalLink>
                    )
                  : (
                      capacityData.source_title
                    )}
                , {capacityData.source_author}
              </p>
            </div>
            <hr />
            <div className="dialog-actions">
              <CapacitySources />
              <Checkbox
                id="show-analytics"
                checked={isVisible}
                onChange={toggleVisible}
                disabled={!isOwnedByCurrentUser()}
              >
                <FormattedMessage
                  id="dialogs.analytics.toggle-visible"
                  defaultMessage="Show capacity counts in segment labels"
                />
              </Checkbox>
              {!isOwnedByCurrentUser() && (
                <p className="analytics-settings-notice">
                  <FontAwesomeIcon icon={ICON_QUESTION_CIRCLE} />
                  <FormattedMessage
                    id="dialogs.analytics.settings-notice"
                    defaultMessage="Street owners can change these settings."
                  />
                </p>
              )}
              <br />
              <Button primary={true} onClick={exportCSV}>
                <FormattedMessage
                  id="dialogs.analytics.export-csv"
                  defaultMessage="Export as CSV"
                />
              </Button>
              <footer>
                <Terms locale={locale} />
              </footer>
            </div>
          </div>
          <button className="dialog-primary-action" onClick={closeDialog}>
            <FormattedMessage id="btn.close" defaultMessage="Close" />
          </button>
        </div>
      )}
    </Dialog>
  )
}

export default AnalyticsDialog
