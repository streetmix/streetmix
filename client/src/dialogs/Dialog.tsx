/**
 * Dialog
 *
 * These things are actually modals, since they block the main window.
 * They're positioned over everything else in the app and will prevent
 * interaction with the rest of the app until it's closed.
 *
 * Clicking on the modal's backdrop will automatically close the modal.
 * We used to have dialogs that prevented this form of closing, but
 * they're not used anymore, so this functionality has been removed.
 * Modals that behave differently from each other is probably not a
 * great idea.
 *
 * Only one modal window is shown at a time. Nested modals aren't supported.
 */
import React, { useState, useEffect, useRef } from 'react'
import { useDispatch } from 'react-redux'
import { CSSTransition } from 'react-transition-group'

import CloseButton from '../ui/CloseButton'
import { useOnClickOutside } from '../ui/useOnClickOutside'
import { clearDialogs } from '../store/slices/dialogs'
import { registerKeypress, deregisterKeypress } from '../app/keypress'
import './Dialog.css'

type CloseFunction = () => void
interface DialogProps {
  children: (arg: CloseFunction) => React.ReactElement
}

function Dialog ({ children }: DialogProps): React.ReactElement {
  // Appear state controls transition in/out
  const dialogEl = useRef(null)
  const nodeRef = useRef(null)
  const [appear, setAppear] = useState(true)
  const dispatch = useDispatch()

  // Set up handler to close dialogs when clicking outside of it
  useOnClickOutside(dialogEl, handleClose)

  // Set up keypress listener to close dialogs if open
  useEffect(() => {
    registerKeypress('esc', handleClose)

    return () => {
      deregisterKeypress('esc', handleClose)
    }
  }, [])

  // On "close", we animate the dialog out
  function handleClose (): void {
    setAppear(false)
  }

  // When the animation is complete, then we clear dialog state
  function handleExit (): void {
    dispatch(clearDialogs())
  }

  return (
    <CSSTransition
      appear={true}
      in={appear}
      timeout={80}
      classNames="dialog-transition"
      onExited={handleExit}
      nodeRef={nodeRef}
    >
      <div className="dialog-box-container" ref={nodeRef}>
        <div className="dialog-box-backdrop" />
        <div className="dialog-box" role="dialog" ref={dialogEl}>
          <CloseButton onClick={handleClose} />
          {children(handleClose)}
        </div>
      </div>
    </CSSTransition>
  )
}

export default Dialog
