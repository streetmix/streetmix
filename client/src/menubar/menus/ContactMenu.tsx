import React from 'react'
import { FormattedMessage } from 'react-intl'

import { useDispatch } from '~/src/store/hooks'
import { showDialog } from '~/src/store/slices/dialogs'
import ExternalLink from '~/src/ui/ExternalLink'
import Icon from '~/src/ui/Icon'
import Menu, { type MenuProps } from './Menu'

function ContactMenu (props: MenuProps): React.ReactElement {
  const dispatch = useDispatch()

  return (
    <Menu {...props}>
      <ExternalLink href="https://strt.mx/discord" icon={true}>
        <Icon name="discord" className="menu-item-icon" />
        <FormattedMessage
          id="menu.contact.discord"
          defaultMessage="Join Discord chat"
        />
      </ExternalLink>
      <ExternalLink href="https://github.com/streetmix/streetmix/" icon={true}>
        <Icon name="github" className="menu-item-icon" />
        <FormattedMessage
          id="menu.contact.github"
          defaultMessage="View source code on GitHub"
        />
      </ExternalLink>
      <a
        onClick={(e) => {
          dispatch(showDialog('NEWSLETTER'))
        }}
      >
        <Icon name="mail" className="menu-item-icon" />
        <FormattedMessage
          id="menu.contact.newsletter"
          defaultMessage="Subscribe to our newsletter"
        />
      </a>
    </Menu>
  )
}

export default ContactMenu
