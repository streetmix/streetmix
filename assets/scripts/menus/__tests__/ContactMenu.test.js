/* eslint-env jest */
import React from 'react'
import { screen } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import { render } from '../../../../test/helpers/render'
import ContactMenu from '../ContactMenu'
import { showDialog } from '../../store/slices/dialogs'

jest.mock('../../store/slices/dialogs', () => ({
  showDialog: jest.fn(() => ({ type: 'MOCK_ACTION' }))
}))

describe('ContactMenu', () => {
  it('renders', () => {
    const { asFragment } = render(<ContactMenu isActive={true} />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('handles clicked menu items', async () => {
    render(<ContactMenu isActive={true} />)

    await userEvent.click(screen.getByText('Discord', { exact: false }))
    await userEvent.click(screen.getByText('GitHub', { exact: false }))
    await userEvent.click(screen.getByText('newsletter', { exact: false }))

    expect(showDialog).toBeCalledTimes(1)
  })
})
