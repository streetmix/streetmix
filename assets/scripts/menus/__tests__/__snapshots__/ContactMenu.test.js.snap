// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactMenu renders 1`] = `
<DocumentFragment>
  <div
    class="menu"
  >
    <a
      href="https://strt.mx/discord"
      rel="noopener noreferrer"
      target="_blank"
    >
      <svg
        class="svg-inline--fa menu-item-icon fa-discord"
      />
      Join Discord chat
    </a>
    <a
      href="https://github.com/streetmix/streetmix/"
      rel="noopener noreferrer"
      target="_blank"
    >
      <svg
        class="svg-inline--fa menu-item-icon fa-github"
      />
      View source code on GitHub
    </a>
    <a>
      Subscribe to our newsletter
    </a>
  </div>
</DocumentFragment>
`;
